M2/Macaulay2/packages/SCMAlgebras.m2

```diff
@@ -0,0 +1,394 @@
+Node
+  Key
+    minDim
+    (minDim,Ideal)
+  Headline
+    computes the minimum dimension of $I$.
+  Usage
+    minDim I
+  Inputs
+    I:Ideal
+      a homogeneous ideal of the polynomial ring $S=K[x_1,\ldots,x_n]$, with $K$ a field
+  Outputs
+    k:ZZ
+      the minimum dimension of $I$
+  Description
+    Text
+      Let $S=K[x_1,\ldots,x_n]$ be a polynomial ring. Given a homogeneous ideal $I\subset S$, it returns the minimum dimension,
+      defined as the minimum integer $k$ such that the filter ideal $I^{<k>}$ differs from $I$.
+    Example
+      S = QQ[x_1..x_10,y_1..y_10];
+      E = {{1,2},{1,3},{1,4},{1,5},{1,6},{1,7},{1,8},{1,9},{1,10},{6,7},{8,9},{8,10},{9,10}};
+      J=ideal(for e in E list x_(e#0)*y_(e#1)-x_(e#1)*y_(e#0));
+      minDim J
+  SeeAlso
+    filterIdeal
+    isSCM
+///);
+-------------------------------------------------------------------------
+minDim = method(TypicalValue=>ZZ);
+minDim(Ideal) := I -> (
+    T:=decompose I;
+    D:=for Q in T list dim(radical Q);
+    return min(D)
+)
+--=======================================================================
+
+
+--=======================================================================
+-- computes the filter ideal I^<i> according to Goodarzi's definition
+--=======================================================================
+MyDoc=concatenate(MyDoc,///
+Node
+  Key
+    filterIdeal
+    (filterIdeal,Ideal,ZZ)
+  Headline
+    computes the $i$th filter ideal of $I$.
+  Usage
+    filterIdeal(I,i)
+  Inputs
+    I:Ideal
+      a homogeneous ideal of the polynomial ring $S=K[x_1,\ldots,x_n]$, with $K$ a field
+    i:ZZ
+      an integer greater or equal than -1
+  Outputs
+    J:Ideal
+      the $i$th filter ideal of $I$
+  Description
+    Text
+      Let $I\subset S$ be a homogeneous ideal, with $d=\dim S/I$, and let $I=\displaystyle\bigcap_{j=1}^r Q_j$ be the minimal primary decomposition of $I$.
+      For all $1\leq j\leq r$, let $P_j = \sqrt{Q_j}$ be the radical of $Q_j$. For all $-1\leq i\leq d$, the $i$th filter ideal of $I$ is $$I^{<i>} = \bigcap_{\dim S/{P_j}>i} Q_{j},$$
+      where $I^{<-1>}=I$ and $I^{<d>}=S$.
+    Example
+      S = QQ[x_1..x_10,y_1..y_10];
+      E = {{1,2},{1,3},{1,4},{1,5},{1,6},{1,7},{1,8},{1,9},{1,10},{6,7},{8,9},{8,10},{9,10}};
+      J=ideal(for e in E list x_(e#0)*y_(e#1)-x_(e#1)*y_(e#0));
+      filterIdeal(J,5)
+  SeeAlso
+    minDim
+    isSCM
+///);
+-------------------------------------------------------------------------
+filterIdeal = method(TypicalValue=>Ideal);
+-------------------------------------------------------------------------
+filterIdeal(Ideal,ZZ) := Ideal => (I,i) -> (
+    S:=ring I;
+    d:=dim I;
+    d0:=minDim I;
+    if i>-2 and i<d0 then (
+        Ii:=I;
+    ) else if i>=d0 and i<d then ( 
+        Q:=decompose I;
+        L:={};
+        for k from 0 to #Q-1 do (
+            Qk:=Q#k;
+            Pk:=radical Qk;
+            dk:=dim Pk;
+            L=L|{{dk,Qk}};
```

Comment:
From the style guide:

> Do not convert an order n algorithm into an order n^2 algorithm by repeated concatenation of lists or of matrices. Similarly, do not repeatedly append or prepend something to a list, and do not repeatedly remove single items from a list. Hash tables are your friend here. Notice also that the for ... list ... command offers a convenient way to make a list.

I think something like `apply(decompose I, J -> {P := radical J, dim P})` would work here.

The same goes for the construction of `Pi` a few lines below.